File: 707.design-linked-list.ts
/*
 * @lc app=leetcode id=707 lang=typescript
 *
 * [707] Design Linked List
 */

// @lc code=start

class LLNode {
  val: number;
  next: LLNode | null;
  constructor(val?: number, next?: LLNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

class MyLinkedList {
  head: LLNode | null;
  constructor(head = null) {
    this.head = head;
  }

  get(index: number): number {
    if (this.head === null) return -1;
    if (index === 0) return this.head.val;
    let temp = this.head.next;
    let count = 0;
    while (count !== index) {
      count++;
      if (temp === null) return -1;
      if (count === index) {
        return temp.val;
      }
      temp = temp.next;
    }
    return -1;
  }

  addAtHead(val: number): void {
    if (this.head === null) {
      this.head = new LLNode(val);
      return;
    }
    let newHead = new LLNode(val);
    let temp = this.head;
    newHead.next = temp;
    this.head = newHead;
  }

  addAtTail(val: number): void {
    if (this.head === null) {
      this.head = new LLNode(val);
      return;
    }
    let temp = this.head;
    while (temp.next !== null) {
      temp = temp.next;
    }
    temp.next = new LLNode(val);
  }

  addAtIndex(index: number, val: number): void {
    if (index === 0) this.addAtHead(val);
    let temp = this.head;
    let count = 0;
    while (count !== index) {
      count++;
      if (temp === null) return;
      if (count === index) {
        let newNode = new LLNode(val);
        newNode.next = temp.next;
        temp.next = newNode;
      }
      temp = temp.next;
    }
    return;
  }

  deleteAtIndex(index: number): void {
    if (this.head === null) return;
    let temp = this.head;
    let count = 0;
    if (index === 0) {
      temp = this.head.next;
      this.head = temp;
      return;
    }
    while (count !== index) {
      count++;
      if (temp === null) return;
      if (count === index) {
        if (temp.next === null) {
          temp = null;
          return;
        }
        temp.next = temp.next.next;
        return;
      }
      temp = temp.next;
    }
    return;
  }
}
/**
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = new MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(val)
 * obj.addAtTail(val)
 * obj.addAtIndex(index,val)
 * obj.deleteAtIndex(index)
 */
// @lc code=end

File: 707.design-linked-list.py
#
# @lc app=leetcode id=707 lang=python3
#
# [707] Design Linked List
#

# @lc code=start
class ListNode:
    def __init__(self, val) -> None:
        self.prev = None
        self.val = val
        self.next = None


class MyLinkedList:

    def __init__(self):
        self.head = None
        self.tail = None

    def get(self, index: int) -> int:
        if not self.head:
            return -1
        i = 0
        curr = self.head
        while i < index:
            if not curr:
                return -1
            curr = curr.next
            i += 1
        return curr.val

    def addAtHead(self, val: int) -> None:
        if not self.head:
            self.head = self.tail = ListNode(val)
            return
        new = ListNode(val)
        new.next = self.head
        self.head.prev = new
        self.head = new

    def addAtTail(self, val: int) -> None:
        if not self.head:
            self.head = self.tail = ListNode(val)
            return
        new = ListNode(val)
        new.prev = self.tail
        self.tail.next = new
        self.tail = new

    def addAtIndex(self, index: int, val: int) -> None:
        i = 0
        if not self.head and index == 0:
            self.head = self.tail = ListNode(val)
            return
        curr = self.head
        while i < index:
            if not curr:
                return
            curr = curr.next
            i += 1
        if not curr:
            new = ListNode(val)
            self.tail.next = new
            new.prev = self.tail
            self.tail = new
            return
        prev = curr.prev
        new = ListNode(val)
        new.prev = prev
        new.next = curr
        if prev:
            prev.next = new
        curr.prev = new
        if curr == self.head:
            self.head = self.head.prev
        if curr == self.tail:
            self.tail = self.tail.next

    def deleteAtIndex(self, index: int) -> None:
        i = 0
        if not self.head:
            return
        curr = self.head
        while i < index:
            if not curr:
                return
            curr = curr.next
            i += 1
        if not curr:
            return
        prev = curr.prev
        next = curr.next
        if prev:
            prev.next = next
        if next:
            next.prev = prev
        if curr == self.head:
            self.head = self.head.next
        if curr == self.tail:
            self.tail = self.tail.prev


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
# @lc code=end
